Major per-RuntimeIdentifier build refactoring

andrew-boyarshin opened the pull request:
* `ExtrasIncludeDefaultProjectBuildOutputInPackTarget` allows overriding the target which packs the RID-specific build output (assemblies and symbols), fixes #217.
* Update to latest `Microsoft.Common.CrossTargeting.targets` upstream source code.
* `_SdkPrepareProjectFlavorMatrix` is now responsible for computing project flavors matrix: `TargetFramework` × `RuntimeIdentifier`.
* Reimplement `_ComputeTargetFrameworkItems` and `_WalkEachTargetPerFramework` on top of `_SdkPrepareProjectFlavorMatrix`
* Mimic `ProjectReference` item spec to simplify `_WalkEachTargetPerFramework` (by removing duplication)
* Use `MSBuildProjectFullPath` consistently, avoid `MSBuildProjectFile` or mixing those two.
* Drop no-op `_ExtrasPackageRuntimeFiles`.
* Pay attention to unnecessary metadata, use `KeepMetadata` where applicable (opt-out possible and verified).

[`PackagePath` override sample](https://github.com/SharpGenTools/SharpGenTools/pull/161/commits/2579a5d1c79ab0c6157dfda4edc3c90958c5393b)

andrew-boyarshin commented:
@clairernovotny it would be great to have this PR reviewed.

clairernovotny commented:
In general looks good. Can you please add a sample/test project that shows the overriding?

Also, what's is the benefit of removing the extra metadata vs the extra complexity that's added? What do we get for it?

andrew-boyarshin commented:
@clairernovotny sure, I'll make a test project and update README to mention the ability to override paths.

My rationale: This project is the foundation of many .NET projects. It should be written like a part of .NET/MSBuild SDK, with practices used there. In particular, the per-RID build feature is so similar to `ProjectReference`s. `ProjectReference`s remove extra metadata to reduce GC footprint. Therefore, _MSBuildSdkExtras_ should do the same.

andrew-boyarshin commented:
@clairernovotny done. I've removed `KeepDuplicates` since the outputs should be unique anyway, no need to add extra `Distinct` call.

clairernovotny commented:
@andrew-boyarshin just lmk when things are ready.

andrew-boyarshin commented:
@clairernovotny fixed and expanded test set (they were using `TargetFramework` instead of `TargetFrameworks`), refactored targets to mimic the original more closely (different source of `TargetFramework` matrix row for different targets), added verification of RID-specific build outputs with MSBuild warning and tests.

Finally ready to be reviewed and merged.

clairernovotny commented:
Thanks!
